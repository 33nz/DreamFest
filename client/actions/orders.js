export const FETCH_ORDERS_PENDING = 'FETCH_ORDERS_PENDING'
export const FETCH_ORDERS_SUCCESS = 'FETCH_ORDERS_SUCCESS'
export const PATCH_ORDER_PENDING = 'PATCH_ORDER_PENDING'
export const PATCH_ORDER_SUCCESS = 'PATCH_ORDER_SUCCESS'

export function fetchOrdersPending () {
  return {
    type: FETCH_ORDERS_PENDING
  }
}

export function fetchOrdersSuccess (orders) {
  return {
    type: FETCH_ORDERS_SUCCESS,
    orders: orders
  }
}

export function patchOrderPending () {
  return {
    type: PATCH_ORDER_PENDING
  }
}

export function patchOrderSuccess (order) {
  return {
    type: PATCH_ORDER_SUCCESS,
    order: order
  }
}
