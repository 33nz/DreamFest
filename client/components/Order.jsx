import React from 'react'
import { Link, Route } from 'react-router-dom'

import OrderDetails from './OrderDetails'

class Order extends React.Component {
  state = {
    order: this.props.order
  }

  updateOrderLine = (id, quantity) => {
    const order = { ...this.state.order }
    const itemToUpdate = order.products.find(item => id === item.id)
    itemToUpdate.quantity = quantity
    this.setState({ order })
  }

  deleteFromOrder = (id) => {
    let order = { ...this.state.order }
    const newProductList = order.products.filter(item => item.id !== id)
    order.products = newProductList
    this.setState({ order })
  }

  render () {
    const { id, createdAt, updatedAt } = this.state.order
    return (
      <div className='order'>
        <p className='name'>Order #{id}</p>
        <p>
          <span>
            <span className='timestamp'>Order placed: {createdAt}</span>
            <br />
            <span className='timestamp'>Last updated: {updatedAt}</span>
          </span>
          <button className='order-button'>
            <Link to={`/orders/${id}`}>View Order</Link>
          </button>
        </p>
        <Route path='/orders/:id' render={(props) => {
          return <OrderDetails
            order={this.state.order}
            updateOrderLine={this.updateOrderLine}
            deleteFromOrder={this.deleteFromOrder}
            match={props.match} />
        }} />
      </div>
    )
  }
}

export default Order
