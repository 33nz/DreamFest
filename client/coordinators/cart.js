export function updateCartItem (updateInfo, updateCartDispatcher) {
  const isValid = !isNaN(Number(updateInfo.quantity))
  isValid && updateCartDispatcher(updateInfo)
}
