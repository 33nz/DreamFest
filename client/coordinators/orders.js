import requestor from '../consume'

export function getOrders (dispatchers, consume = requestor) {
  const { fetchOrdersPending, fetchOrdersSuccess, showError } = dispatchers
  fetchOrdersPending()
  return consume('/orders')
    .then(res => {
      const orders = res.body
      fetchOrdersSuccess(orders)
      return null
    })
    .catch(err => {
      showError(err.message)
    })
}
