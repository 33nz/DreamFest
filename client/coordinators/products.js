import requestor from '../consume'

export function getProducts (dispatchers, consume = requestor) {
  const { fetchProductsPending, fetchProductsSuccess, showError } = dispatchers
  fetchProductsPending()
  return consume('/products')
    .then(res => {
      const products = res.body
      fetchProductsSuccess(products)
      return null
    })
    .catch(err => {
      showError(err.message)
    })
}
