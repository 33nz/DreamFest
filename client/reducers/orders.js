import {
  FETCH_ORDERS_SUCCESS,
  POST_ORDER_SUCCESS,
  PATCH_ORDER_SUCCESS,
  DELETE_ORDER_SUCCESS
} from '../actions/orders'

const orders = (state = [], action) => {
  switch (action.type) {
    case FETCH_ORDERS_SUCCESS:
      return action.orders

    case POST_ORDER_SUCCESS:
      return [ ...state, action.order ]

    case PATCH_ORDER_SUCCESS:
      return state.map(ord => (ord.id === action.order.id) ? action.order : ord)

    case DELETE_ORDER_SUCCESS:
      return state.filter(order => order.id !== action.orderId)

    default:
      return state
  }
}

export default orders
