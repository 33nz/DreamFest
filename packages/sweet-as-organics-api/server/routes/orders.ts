import express from 'express'
import * as db from '../db/orders'

const router = express.Router()

export default router

router.get('/', (req, res) => {
  db.listOrders()
    .then((orders) => {
      console.log(orders)
      res.json(orders)
    })
    .catch((err) => {
      console.log(err)
      res.status(500).send('DATABASE ERROR: ' + err.message)
    })
})

router.post('/', (req, res) => {
  const { cart } = req.body

  if (!Array.isArray(cart)) {
    return res.status(400).send('INVALID ORDER: cart must be an array')
  }

  db.addOrder(cart)
    .then(() => {
      res.sendStatus(201)
    })
    .catch((err) => {
      res.status(500).send('DATABASE ERROR: ' + err.message)
    })
})

router.patch('/:id', (req, res) => {
  const id = Number(req.params.id)
  if (isNaN(id)) return res.sendStatus(400)
  const { status } = req.body

  const validStatuses = ['pending', 'cancelled', 'completed']
  if (!validStatuses.includes(status)) {
    return res
      .status(400)
      .send('INPUT ERROR: status must be one of ' + validStatuses.join(', '))
  }

  db.editOrderStatus(id, status)
    .then((updatedOrder) => {
      res.json(updatedOrder)
    })
    .catch((err) => {
      res.status(500).send('DATABASE ERROR: ' + err.message)
    })
})
