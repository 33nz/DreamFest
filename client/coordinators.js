import { showError } from './actions/error'
import { requestProducts, receiveProducts } from './actions/products'

import { fetchProducts } from './api'

export function getProducts (dispatch) {
  dispatch(requestProducts())
  fetchProducts()
    .then(products => {
      dispatch(receiveProducts(products))
    })
    .catch(err => {
      dispatch(showError(err.message))
    })
}
