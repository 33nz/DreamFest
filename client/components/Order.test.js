import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'

import { Order } from './Order'

import mockOrders from '../testing/mockOrders'

import { updateOrder } from '../coordinators/orders'
jest.mock('../coordinators/orders', () => {
  return {
    updateOrder: jest.fn()
  }
})

test('renders pending order correctly', () => {
  const order = mockOrders[1]
  const { asFragment } = render(<Order order={order} />)

  expect(asFragment()).toMatchSnapshot()
})

test('renders cancelled order correctly', () => {
  const order = mockOrders[0]
  const { asFragment } = render(<Order order={order} />)

  expect(asFragment()).toMatchSnapshot()
})

test('calls updateOrder on cancel button click', async () => {
  const order = mockOrders[1]
  render(<Order order={order} />)

  const cancelButton = await screen.getByRole('button', { name: 'Cancel Order' })
  fireEvent.click(cancelButton)

  expect(updateOrder).toHaveBeenCalled()
})
