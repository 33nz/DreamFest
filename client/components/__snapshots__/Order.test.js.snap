// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders cancelled order correctly 1`] = `
<DocumentFragment>
  <div
    class="order"
  >
    <p
      class="name"
    >
      Order #1
    </p>
    <p
      class="order-details"
    >
      Order placed: 10:32:45 AM, Thu Sep 10 2020
    </p>
    <p
      class="order-details"
    >
      <span
        aria-hidden="true"
        class="fa fa-circle cancelled"
      />
      Status: cancelled
    </p>
    <table>
      <thead>
        <tr>
          <td
            role="columnheader"
          >
            Product
          </td>
          <td
            role="columnheader"
          >
            Quantity
          </td>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            snazzy product item
          </td>
          <td>
            3
          </td>
        </tr>
      </tbody>
    </table>
    <div />
  </div>
</DocumentFragment>
`;

exports[`renders pending order correctly 1`] = `
<DocumentFragment>
  <div
    class="order"
  >
    <p
      class="name"
    >
      Order #2
    </p>
    <p
      class="order-details"
    >
      Order placed: 10:39:26 AM, Thu Sep 10 2020
    </p>
    <p
      class="order-details"
    >
      <span
        aria-hidden="true"
        class="fa fa-circle pending"
      />
      Status: pending
    </p>
    <table>
      <thead>
        <tr>
          <td
            role="columnheader"
          >
            Product
          </td>
          <td
            role="columnheader"
          >
            Quantity
          </td>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            snazzy product item
          </td>
          <td>
            2
          </td>
        </tr>
        <tr>
          <td>
            snazzier product item
          </td>
          <td>
            1
          </td>
        </tr>
      </tbody>
    </table>
    <div>
      <button
        class="order-button"
      >
        Cancel Order
      </button>
    </div>
  </div>
</DocumentFragment>
`;
